be/src/util/frame_of_reference_coding.cpp

```diff
@@ -0,0 +1,377 @@
+        _buffered_values_num = ForCoding::FRAME_VALUE_NUM;
+        bit_packing_one_frame_value(in_data);
+        in_data += ForCoding::FRAME_VALUE_NUM;
+    }
+
+    // 3. process remaining value
+    size_t remaining_num = (count - padding_num) % ForCoding::FRAME_VALUE_NUM;
+    if (remaining_num > 0) {
+        copy_value(in_data, remaining_num);
+    }
+
+    _count += count;
+}
+
+// todo(kks): improve this method by SIMD instructions
+
+// Use as few bit as possible to store a piece of integer data.
+// param[in] input: the integer list need to pack
+// param[in] in_num: the number integer need to pack
+// param[in] bit_width: how many bit we use to store each integer data
+// param[out] out: the packed result
+
+// For example:
+// The input is int32 list: 1, 2, 4, 8 and bit_width is 3
+// The output will be: 0001 0010 0100 1000
+template<typename T>
+void ForEncoder<T>::bit_pack(T *input, uint8_t in_num, int bit_width, uint8_t *output) {
+    T in_mask = 0;
+    int bit_index = 0;
+    *output = 0;
+    for (int i = 0; i < in_num; i++) {
+        in_mask = 1 << (bit_width - 1);
+        for (int k = 0; k < bit_width; k++) {
+            if (bit_index > 7) {
+                bit_index = 0;
+                output++;
+                *output = 0;
+            }
+            *output |= (((input[i] & in_mask) >> (bit_width - k - 1)) << (7 - bit_index));
+            in_mask >>= 1;
+            bit_index++;
+        }
+    }
+}
+
+template<typename T>
+void ForEncoder<T>::bit_packing_one_frame_value(const T* input) {
+    T min = input[0];
+    T max = input[0];
+    bool is_ascending = true;
+    uint8_t bit_width = 0;
+
+    for (uint8_t i = 1; i < _buffered_values_num; ++i) {
+        if (is_ascending) {
+            if (input[i] < input[i - 1]) {
+                is_ascending = false;
+            } else {
+                bit_width = std::max(bit_width, bits(input[i]- input[i - 1])) ;
+            }
+        }
+
+        if (input[i] < min) {
+            min = input[i];
+            continue;
+        }
+
+        if (input[i] > max) {
+            max = input[i];
+        }
+    }
+
+    if (sizeof(T) == 8) {
+        put_fixed64_le(_buffer, min);
+    } else {
+        put_fixed32_le(_buffer, min);
+    }
+
+    T delta_values[_buffered_values_num];
+    // improve for ascending order input, we could use fewer bit
+    if (is_ascending) {
+        delta_values[0] = 0;
+        for (uint8_t i = 1; i < _buffered_values_num; ++i) {
+            delta_values[i] = input[i] - input[i - 1];
+        }
+        _order_flags.push_back(true);
+    } else {
+        bit_width = bits(static_cast<T>(max - min));
+        for (uint8_t i = 0; i < _buffered_values_num; ++i) {
+            delta_values[i] = input[i] - min;
+        }
+        _order_flags.push_back(false);
+    }
+    _bit_widths.push_back(bit_width);
+
+    uint32_t packing_len = BitUtil::Ceil(_buffered_values_num * bit_width, 8);
+    uint8_t packing_values[packing_len];
```

Comment:
another VLA usage. You can resize `_buffer` first and write packed bytes directly into `_buffer->data() + _buffer->size()`